### ISSUE TITLE
why epoch-Train-perplexity is nan in the mxnet lstm training, but batch-Train-perplexity is normal and keep decline?

### ISSUE THREAD

**Issue by whaozl** [apache/incubator-mxnet#4916]
in mxnet 0.9.3

![476774188 chatroom_1486442908967_50](https://cloud.githubusercontent.com/assets/6327360/22678394/ad02424a-ed34-11e6-8341-a6d8f44c5a38.png)

use perplexity function:
```python
def perplexity(label, pred):
    label = label.T.reshape((-1,)) 
    loss = 0. 
    for i in range(pred.shape[0]): 
        loss += -np.log(max(1e-10, pred[i][int(label[i])])) 
    return np.exp(loss / label.size)
```

**But I don't see the Resetting Data Iterator process. Why?**

**Comment by angelsrates** [comment]
Hi there,
Did you figure this out? I am getting the same issue. Thanks!

**Comment by JACKHAHA363** [comment]
Here is my guess:
I dig into the code for Speedometer, and there is a thrid parameter auto-reset which is true by default. It seems it would reset your metric when you call this callback. Then your number of instances becomes zero, and if that is exactly the end of the epoch, when you get the metric value, you would divide by 0, which returns NaN.

So, are you using auto-reset by default?

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
